go/src/koding/kites/kloud/provider/do/machine.go

```diff
@@ -0,0 +1,214 @@
+	dropletID, err := m.DropletID()
+	if err != nil {
+		return nil, err
+	}
+
+	if dropletID == 0 {
+		return nil, ErrInvalidDropletID
+	}
+
+	action, _, err := m.client.DropletActions.PowerOn(dropletID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := waitForAction(ctx, m.client, action); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// Stop stops an existing droplet associated to the machine
+func (m *Machine) Stop(ctx context.Context) (interface{}, error) {
+	dropletID, err := m.DropletID()
+	if err != nil {
+		return nil, err
+	}
+
+	if dropletID == 0 {
+		return nil, ErrInvalidDropletID
+	}
+
+	// The preferred way to turn off a Droplet is to attempt a shutdown, with a
+	// reasonable timeout, followed by a power off action to ensure the Droplet
+	// is off.
+	action, _, err := m.client.DropletActions.Shutdown(dropletID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := waitForAction(ctx, m.client, action); err != nil {
+		return nil, err
+	}
+
+	action, _, err = m.client.DropletActions.PowerOff(dropletID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := waitForAction(ctx, m.client, action); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// Info returns the droplet state
+func (m *Machine) Info(context.Context) (machinestate.State, interface{}, error) {
+	dropletID, err := m.DropletID()
+	if err != nil {
+		return machinestate.Unknown, nil, err
+	}
+
+	if dropletID == 0 {
+		return machinestate.Unknown, nil, ErrInvalidDropletID
+	}
+
+	droplet, resp, err := m.client.Droplets.Get(dropletID)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			return machinestate.NotInitialized, nil, nil
+		}
+
+		return machinestate.Unknown, nil, err
+	}
+
+	return statusToState(droplet.Status), nil, nil
+}
+
+// DropletID returns the droplet ID associated for this given machine
+func (m *Machine) DropletID() (int, error) {
+	metadata, ok := m.BaseMachine.Metadata.(*Metadata)
+	if !ok {
+		return 0, fmt.Errorf("meta data is not of type do.Metadata: %T",
+			m.BaseMachine.Metadata)
+	}
+	return metadata.DropletID, nil
+}
+
+// waitForAction waits for a single action to finish.
+func waitForAction(ctx context.Context, client *godo.Client, action *godo.Action) error {
+	if action == nil {
+		return nil
+	}
+
+	// NOTE: if we need debugging, enable the following
+	debug := false
+	if debug {
+		start := time.Now()
+		log.Println("waiting for action to finish: ", action.ID)
+		defer func() {
+			log.Println("done: ", time.Since(start).Seconds())
+		}()
+	}
+
+	ticker := backoff.NewTicker(backoff.NewExponentialBackOff())
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			var err error
+			action, _, err = client.Actions.Get(action.ID)
+			if err != nil {
+				return err
+			}
+			if action.Status == actionErrored {
+				return errors.New(action.String())
+			}
+			if action.CompletedAt != nil || action.Status == actionCompleted {
+				return nil
+			}
+		case <-ctx.Done():
```

Comment:
No we don't, moreover context are used wrongly - for storing global-scoped data. I'm in ongoing effort to have that all fixed.